fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biomes/MixinSetBaseBiomesLayer.java

```diff
@@ -0,0 +1,150 @@
+		if (moddedWeightTotal == null) {
+			return;
+		}
+		int vanillaArrayWeight = vanillaArray.length;
+		double reqWeightSum = (double) random.nextInt(Integer.MAX_VALUE) * (vanillaArray.length + moddedWeightTotal) / Integer.MAX_VALUE;
+		if (reqWeightSum < vanillaArrayWeight) {
+			info.setReturnValue(vanillaArray[(int) reqWeightSum]);
+		}
+		else {
+			List<BaseBiomeEntry> moddedBiomes = InternalBiomeData.getOverworldModdedBaseBiomes().get(climate);
+			info.setReturnValue(moddedBiomes.get(InternalBiomeUtils.searchForBiome(reqWeightSum, vanillaArrayWeight, moddedBiomes)).getRawId());
+		}
+	}
+
+	@Inject(at = @At(value = "FIELD", target = "Lnet/minecraft/world/biome/layer/SetBaseBiomesLayer;SNOWY_BIOMES:[I"), method = "sample", cancellable = true)
+	private void injectSnowyBiomes(LayerRandomnessSource random, int value, CallbackInfoReturnable<Integer> info) {
+		int[] vanillaArray = SNOWY_BIOMES;
+		OverworldClimate climate = OverworldClimate.SNOWY;
+		Double moddedWeightTotal = InternalBiomeData.getOverworldModdedBaseBiomeWeightTotals().get(climate);
+		if (moddedWeightTotal == null) {
+			return;
+		}
+		int vanillaArrayWeight = vanillaArray.length;
+		double reqWeightSum = (double) random.nextInt(Integer.MAX_VALUE) * (vanillaArray.length + moddedWeightTotal) / Integer.MAX_VALUE;
+		if (reqWeightSum < vanillaArrayWeight) {
+			info.setReturnValue(vanillaArray[(int) reqWeightSum]);
+		}
+		else {
+			List<BaseBiomeEntry> moddedBiomes = InternalBiomeData.getOverworldModdedBaseBiomes().get(climate);
+			info.setReturnValue(moddedBiomes.get(InternalBiomeUtils.searchForBiome(reqWeightSum, vanillaArrayWeight, moddedBiomes)).getRawId());
+		}
+	}
+
+	@Inject(at = @At(value = "FIELD", target = "Lnet/minecraft/world/biome/layer/SetBaseBiomesLayer;COOL_BIOMES:[I"), method = "sample", cancellable = true)
+	private void injectCoolBiomes(LayerRandomnessSource random, int value, CallbackInfoReturnable<Integer> info) {
+		int[] vanillaArray = COOL_BIOMES;
+		OverworldClimate climate = OverworldClimate.COOL;
+		Double moddedWeightTotal = InternalBiomeData.getOverworldModdedBaseBiomeWeightTotals().get(climate);
+		if (moddedWeightTotal == null) {
+			return;
+		}
+		int vanillaArrayWeight = vanillaArray.length;
+		double reqWeightSum = (double) random.nextInt(Integer.MAX_VALUE) * (vanillaArray.length + moddedWeightTotal) / Integer.MAX_VALUE;
+		if (reqWeightSum < vanillaArrayWeight) {
+			info.setReturnValue(vanillaArray[(int) reqWeightSum]);
+		}
+		else {
+			List<BaseBiomeEntry> moddedBiomes = InternalBiomeData.getOverworldModdedBaseBiomes().get(climate);
+			info.setReturnValue(moddedBiomes.get(InternalBiomeUtils.searchForBiome(reqWeightSum, vanillaArrayWeight, moddedBiomes)).getRawId());
+		}
+	}
+
+	@Inject(at = @At("RETURN"), method = "sample", cancellable = true)
+	private void transformVariants(LayerRandomnessSource random, int value, CallbackInfoReturnable<Integer> info) {
+		Biome biome = Registry.BIOME.get(value);
```

Comment:
Found a bug when testing Terrestria, this caused my biome variants to fail - and likely caused random biomes to be errornously replaced with "variants" for other mods as well
```suggestion
		Biome biome = Registry.BIOME.get(info.getReturnValue());
```